cliext/oauth.go

```diff
@@ -0,0 +1,244 @@
+package cliext
+
+import (
+	"context"
+	"crypto/rand"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/pkg/browser"
+	"golang.org/x/oauth2"
+)
+
+type OAuthClientConfig struct {
+	// ClientID is the OAuth client ID.
+	ClientID string
+	// ClientSecret is the OAuth client secret.
+	ClientSecret string
+	// TokenURL is the OAuth token endpoint URL.
+	TokenURL string
+	// AuthURL is the OAuth authorization endpoint URL.
+	AuthURL string
+	// Scopes are the requested OAuth scopes.
+	Scopes []string
+	// RequestParams are additional parameters to include in OAuth requests.
+	RequestParams map[string]string
+}
+
+type OAuthToken struct {
+	// AccessToken is the current access token.
+	AccessToken string
+	// AccessTokenExpiresAt is when the access token expires.
+	AccessTokenExpiresAt time.Time
+	// AccessTokenRefreshed indicates whether the access token was just refreshed.
+	AccessTokenRefreshed bool
+	// RefreshToken is the refresh token for obtaining new access tokens.
+	RefreshToken string
+	// TokenType is the type of token (usually "Bearer").
+	TokenType string
+}
+
+type OAuthConfig struct {
+	OAuthClientConfig
+	OAuthToken
+}
+
+func oauthTokenFromOAuth2(token *oauth2.Token) OAuthToken {
+	return OAuthToken{
+		AccessToken:          token.AccessToken,
+		AccessTokenExpiresAt: token.Expiry,
+		RefreshToken:         token.RefreshToken,
+		TokenType:            token.TokenType,
+	}
+}
+
+// OAuthClient handles OAuth authentication using the Authorization Code Flow.
+// See https://tools.ietf.org/html/rfc6749#section-4.1 for details.
+type OAuthClient struct {
+	Options OAuthClientConfig
+	// OnAuthURL is called with the authorization URL. If nil, the URL is opened in the browser.
+	// This can be set to override the default behavior, e.g., for testing.
+	OnAuthURL func(authURL string)
+}
+
+func NewOAuthClient(opts OAuthClientConfig) *OAuthClient {
+	return &OAuthClient{Options: opts}
+}
+
+func (c *OAuthClient) createOAuth2Config(redirectURL string) *oauth2.Config {
+	return &oauth2.Config{
+		ClientID:     c.Options.ClientID,
+		ClientSecret: c.Options.ClientSecret,
+		RedirectURL:  redirectURL,
+		Endpoint: oauth2.Endpoint{
+			AuthURL:  c.Options.AuthURL,
+			TokenURL: c.Options.TokenURL,
+		},
+		Scopes: c.Options.Scopes,
+	}
+}
+
+type oauthCallbackResult struct {
+	code string
+	err  error
+}
+
+// Login performs the OAuth Authorization Code Flow with PKCE (Proof Key for Code Exchange).
+//
+// It starts a local HTTP server to receive the callback, generates an authorization URL,
+// and exchanges the authorization code for a token.
+func (c *OAuthClient) Login(ctx context.Context) (OAuthToken, error) {
```

Comment:
I would not expect this to be part of `cliext`, it is only for cloud CLI right? I think the extension should be in charge of all of this code, it is not code that will ever be used by the CLI or any other extension I would assume nor should you want to maintain it over here IMO.